### PR TITLE
fetchLeverageTiers

### PR DESCRIPTION
**Pull request by samgermain**
I was going to just rewrite `loadLeverageBrackets` but it looks like you might use the array structure internally for stuff

# Ascendex

```
2022-02-10T04:49:52.671Z
Node.js: v14.17.0
CCXT v1.72.43
ascendex.fetchLeverageTiers (ADA/USDT:USDT)
0 ms
{
  'ADA/USDT:USDT': [
    {
      tier: 0,
      notionalFloor: 0,
      notionalCap: 5000,
      maintenanceMarginRatio: 0.006,
      maxLeverage: 100,
      info: {
        positionNotionalLowerBound: '0',
        positionNotionalUpperBound: '5000',
        initialMarginRate: '0.01',
        maintenanceMarginRate: '0.006'
      }
    },
    {
      tier: 1,
      notionalFloor: 5000,
      notionalCap: 20000,
      maintenanceMarginRatio: 0.012,
      maxLeverage: 50,
      info: {
        positionNotionalLowerBound: '5000',
        positionNotionalUpperBound: '20000',
        initialMarginRate: '0.02',
        maintenanceMarginRate: '0.012'
      }
    },
    ...
    {
      tier: 5,
      notionalFloor: 4000000,
      notionalCap: 1000000000,
      maintenanceMarginRatio: 0.2,
      maxLeverage: 3.000003000003,
      info: {
        positionNotionalLowerBound: '4000000',
        positionNotionalUpperBound: '1000000000',
        initialMarginRate: '0.333333',
        maintenanceMarginRate: '0.2'
      }
    }
  ]
}
```

# Binance

```
2022-02-05T07:24:11.251Z
Node.js: v14.17.0
CCXT v1.72.29
binanceusdm.fetchLeverageTiers ()
185 ms
{
  'RAY/USDT': [
    {
      tier: 1,
      notionalFloor: 0,
      notionalCap: 5000,
      maintenanceMarginRatio: 0.01,
      maxLeverage: 25,
      info: {
        bracket: '1',
        initialLeverage: '25',
        notionalCap: '5000',
        notionalFloor: '0',
        maintMarginRatio: '0.01',
        cum: '0.0'
      }
    },
    ...
    {
      tier: 6,
      notionalFloor: 1000000,
      notionalCap: 30000000,
      maintenanceMarginRatio: 0.5,
      maxLeverage: 1,
      info: {
        bracket: '6',
        initialLeverage: '1',
        notionalCap: '30000000',
        notionalFloor: '1000000',
        maintMarginRatio: '0.5',
        cum: '386950.0'
      }
    }
  ]
}
```

# kucoinfutures

```
2022-02-05T07:25:02.845Z
Node.js: v14.17.0
CCXT v1.72.29
kucoinfutures.fetchLeverageTiers (ETH/USDT:USDT)
264 ms
{
  'ETH/USDT:USDT': [
    {
      tier: 1,
      notionalFloor: 0,
      notionalCap: 300000,
      maintenanceMarginRatio: 0.005,
      maxLeverage: 100,
      info: {
        symbol: 'ETHUSDTM',
        level: 1,
        maxRiskLimit: 300000,
        minRiskLimit: 0,
        maxLeverage: 100,
        initialMargin: 0.01,
        maintainMargin: 0.005
      }
    },
    {
      tier: 2,
      notionalFloor: 300000,
      notionalCap: 700000,
      maintenanceMarginRatio: 0.007,
      maxLeverage: 75,
      info: {
        symbol: 'ETHUSDTM',
        level: 2,
        maxRiskLimit: 700000,
        minRiskLimit: 300000,
        maxLeverage: 75,
        initialMargin: 0.013,
        maintainMargin: 0.007
      }
    },
    ...
  ]
}
```

# okx
```
2022-02-05T07:25:23.773Z
Node.js: v14.17.0
CCXT v1.72.29
okx.fetchLeverageTiers (ETH/USDT:USDT)
306 ms
{
  'ETH/USDT:USDT': [
    {
      tier: 1,
      notionalFloor: 0,
      notionalCap: 3000,
      maintenanceMarginRatio: 0.004,
      maxLeverage: 125,
      info: {
        baseMaxLoan: '',
        imr: '0.008',
        instId: '',
        maxLever: '125',
        maxSz: '3000',
        minSz: '0',
        mmr: '0.004',
        optMgnFactor: '0',
        quoteMaxLoan: '',
        tier: '1',
        uly: 'ETH-USDT'
      }
    },
    {
      tier: 2,
      notionalFloor: 3001,
      notionalCap: 6000,
      maintenanceMarginRatio: 0.005,
      maxLeverage: 100,
      info: {
        baseMaxLoan: '',
        imr: '0.01',
        instId: '',
        maxLever: '100',
        maxSz: '6000',
        minSz: '3001',
        mmr: '0.005',
        optMgnFactor: '0',
        quoteMaxLoan: '',
        tier: '2',
        uly: 'ETH-USDT'
      }
    },
    ...
  ]
}
```

# Xena

```
2022-02-10T04:51:27.472Z
Node.js: v14.17.0
CCXT v1.72.43
xena.fetchLeverageTiers (BTC/USD:USDC)
1 ms
{
  'BTC/USD:USDC': [
    {
      tier: 0,
      notionalFloor: 0,
      notionalCap: 10,
      maintenanceMarginRatio: 0.025,
      maxLeverage: 20,
      info: {
        maxVolume: '10',
        initialRate: '0.05',
        maintenanceRate: '0.025'
      }
    },
    {
      tier: 1,
      notionalFloor: 10,
      notionalCap: 20,
      maintenanceMarginRatio: 0.05,
      maxLeverage: 10,
      info: { maxVolume: '20', initialRate: '0.1', maintenanceRate: '0.05' }
    },
    {
      tier: 2,
      notionalFloor: 20,
      notionalCap: 30,
      maintenanceMarginRatio: 0.1,
      maxLeverage: 5,
      info: { maxVolume: '30', initialRate: '0.2', maintenanceRate: '0.1' }
    },
    ...
    {
      tier: 6,
      notionalFloor: 150,
      notionalCap: 200,
      maintenanceMarginRatio: 0.5,
      maxLeverage: 1,
      info: { maxVolume: '200', initialRate: '1', maintenanceRate: '0.5' }
    }
  ]
}
```

### CODE REVIEW

**Comment by frosty00** [comment]
I don't think we should keep adding non-unified stuff to ccxt, for example your last commits include:


https://github.com/ccxt/ccxt/pull/11344
https://github.com/ccxt/ccxt/pull/11345

^ none of these methods are unified. Adding new methods to ccxt makes the unification process harder. I'd suggest to avoid doing it unless you can add the same method to more than one exchange

**Comment by frosty00** [comment]
would appreciate if you can continue to work on adding fetchFundingRate / fetchFundingHistory where it is missing

**Comment by samgermain** [comment]
#11344 was needed to solve this issue
#11345 is something I need for freqtrade

**Comment by samgermain** [comment]
> ^ none of these methods are unified. Adding new methods to ccxt makes the unification process harder. I'd suggest to avoid doing it unless you can add the same method to more than one exchange

Also, if I create a method for multiple exchanges before opening a PR, that's a lot more wasted time if the PR get's closed before it's merged. I can always add on more exchanges after opening it before it get's merged

**Comment by frosty00** [comment]
> Also, if I create a method for multiple exchanges before opening a PR, that's a lot more wasted time if the PR get's closed before it's merged. I can always add on more exchanges after opening it before it get's merged

The point is that there's like 20 methods or so that have missing implementations that we can focus on before trying to add new stuff.

**Comment by kroitor** [comment]
@samgermain 

> Also, if I create a method for multiple exchanges before opening a PR, that's a lot more wasted time if the PR get's closed before it's merged. I can always add on more exchanges after opening it before it get's merged

We must have more than implementation in sight to be able to unify otherwise a single-exchange implementation is always exchange-specific. We cannot add a method just to end up completely rewriting it later when we encounter another exchange having a different structure, different naming and different requirements for the same data.

**Comment by samgermain** [comment]
@kroitor How do you feel about adding a field `maintenanceMarginRate` to markets, it looks like most exchanges that don't have something like this, just have a static value for maintenance margin returned in markets

**Comment by kroitor** [comment]
> @kroitor How do you feel about adding a field `maintenanceMarginRate` to markets, it looks like most exchanges that don't have something like this, just have a static value for maintenance margin returned in markets

I'm fine with that.
@frosty00 do you have any objections?

**Comment by samgermain** [comment]
@kroitor Should be good, I wrote it for bybit also, but it needed some work so I reverted it for now

File: js/kucoinfutures.js [hunk 1]
```diff
@@ -1594,4 +1596,54 @@ module.exports = class kucoinfutures extends kucoin {
...
+        for (let i = 0; i < data.length; i++) {
+            const tier = data[i];
+            const symbol = this.safeSymbol (this.safeString (tier, 'symbol'));
+            if (!(symbol in tiers)) {
+                tiers[symbol] = [];
+            }
+            tiers[symbol].push ({
+                'tier': this.safeNumber (tier, 'level'),
+                'notionalFloor': this.safeNumber (tier, 'minRiskLimit'),
+                'notionalCap': this.safeNumber (tier, 'maxRiskLimit'),
```
Comment (samgermain) [inline js/kucoinfutures.js:1641]: This is the only part I'm unsure about, I'm not sure if the `minRiskLimit` and `maxRiskLimit` are actually the `notionalFloor` and `notionalCap`

**Comment by samgermain** [comment]
I guess I'll need to add tests for this PR, I'll do that, and add some documentation as well

**Comment by samgermain** [comment]
I was actually able to get [multiple tiers for mexc based on this](https://support.mexc.com/hc/article_attachments/360059033152/mceclip0.png)

**Comment by samgermain** [comment]
@kroitor While loops aren't allowed?

**Comment by kroitor** [comment]
> @kroitor While loops aren't allowed?

Nope just for-loops.

**Comment by samgermain** [comment]
I think I can add the leverage tiers for aax, gateio, and delta as well, but there's some things I don't know how to figure out right now. 

I think this PR could be merged, and I could probably figure out those 3 exchanges on separate PRs

**Comment by samgermain** [comment]
Dividing into separate PRs

**Pull request closed without merging**